domain/keyupdater/state/controllerstate.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/model"
+	"github.com/juju/juju/domain"
+	modelerrors "github.com/juju/juju/domain/model/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// ControllerState provides a state access layer for accessing a controller's
+// ssh keys via controller config.
+type ControllerState struct {
+	*domain.StateBase
+}
+
+// GetControllerConfigKeys returns the controller config key and values for the
+// keys supplied. If one or more keys supplied do not exist in the controller's
+// config they will be omitted from the final result.
+func (st *ControllerState) GetControllerConfigKeys(
+	ctx context.Context,
+	keys []string,
+) (map[string]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Errorf(
+			"getting database when getting controller config keys: %w", err,
+		)
+	}
+
+	sqlKeys := make(sqlair.S, 0, len(keys))
+	for _, key := range keys {
+		sqlKeys = append(sqlKeys, key)
+	}
+
+	stmt, err := st.Prepare(`
+SELECT &keyValue.*
+FROM v_controller_config
+WHERE key IN ($S[:])
+`, keyValue{}, sqlKeys)
+	if err != nil {
+		return nil, errors.Errorf(
+			"preparing statement for getting keys from controller config: %w",
+			err,
+		)
+	}
+
+	keyValues := []keyValue{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, sqlKeys).GetAll(&keyValues)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return err
```

Comment:
Once we agree on how to Trace(), we'll want that here. Not sure if we want to stick with errors.Trace(err) until we get there.

Reply:
I don't think it is needed in this context as it is annotated below.